source/extensions/filters/http/cache/cache_header_utility.cc

```diff
@@ -0,0 +1,107 @@
+#include "extensions/filters/http/cache/cache_header_utility.h"
+
+#include <vector>
+
+#include "common/common/utility.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/header_utility.h"
+#include "common/http/headers.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+namespace {
+const absl::string_view bytes_{"bytes"};
+}
+
+std::vector<RawByteRange> CacheHeaderUtility::getRanges(const Http::HeaderMap& request_headers) {
+  if (request_headers.Method() == nullptr ||
+      request_headers.Method()->value().getStringView() != Http::Headers::get().MethodValues.Get) {
+    // Range headers are only valid on GET requests so don't bother parsing
+    // range if we're not going to use it.
+    return {};
+  }
+
+  // Multiple instances of range/range-unit headers are considered invalid.
+  // https://tools.ietf.org/html/rfc7230#section-3.2.2
+  // TODO: should we log when encountering multiple headers?
+  std::vector<absl::string_view> range_unit_headers;
+  Http::HeaderUtility::getAllOfHeader(request_headers, Http::Headers::get().RangeUnit.get(),
+                                      range_unit_headers);
+  absl::string_view range_unit;
+  if (range_unit_headers.size() == 1) {
+    range_unit = range_unit_headers.front();
+  } else if (range_unit_headers.empty()) {
+    range_unit = bytes_;
+  } else {
+    return {};
+  }
+
+  std::vector<absl::string_view> range_headers;
+  Http::HeaderUtility::getAllOfHeader(request_headers, Http::Headers::get().Range.get(),
+                                      range_headers);
+  absl::string_view range;
+  if (range_headers.size() == 1) {
+    range = range_headers.front();
+  } else {
+    return {};
+  }
+
+  return parseRangeHeaderValue(range_unit, range);
+}
+
+std::vector<RawByteRange> CacheHeaderUtility::parseRangeHeaderValue(absl::string_view range_unit,
+                                                               absl::string_view range) {
+  if (range_unit.empty() || !absl::ConsumePrefix(&range, range_unit) ||
+      !absl::ConsumePrefix(&range, "=")) {
+    return {};
+  }
+
+  std::vector<RawByteRange> ranges;
+  while (!range.empty()) {
+    uint64_t first, last;
+    if (!StringUtil::consumeLeadingDigits(&range, &first)) {
```

Comment:
I have an idea that I think can make the body of this while loop easier to read:
if (!StringUtil::consumeLeadingDigits(&range, &first)) {
  first = UINT64_MAX;
}
if (!absl::ConsumePrefix(&range, "-")) {
      ranges.clear();
      break;
}
if (!StringUtil::consumeLeadingDigits(&range, &last)) {
  last = UINT64_MAX;
}
if (last < first) {
      ranges.clear();
      break;
}

Unteseted, so I may be making obvious mistakes.

Reply:
It seems incrementally better after extending the above to handle all edge cases:

```c++ 48   std::vector<RawByteRange> ranges;                                                                                                                                                                                                                       
 49   while (!range.empty()) {                                                                                                                                                                                                                                
 50     absl::optional<uint64_t> first = StringUtil::readAndRemoveLeadingDigits(range);                                                                                                                                                                       
 51     if (!first) {                                                                                                                                                                                                                                         
 52       first = UINT64_MAX;                                                                                                                                                                                                                                 
 53     }                                                                                                                                                                                                                                                     
 54     if (!absl::ConsumePrefix(&range, "-")) {                                                                                                                                                                                                              
 55       ranges.clear();                                                                                                                                                                                                                                     
 56       break;                                                                                                                                                                                                                                              
 57     }                                                                                                                                                                                                                                                     
 58     absl::optional<uint64_t> last = StringUtil::readAndRemoveLeadingDigits(range);                                                                                                                                                                        
 59     if (!last) {                                                                                                                                                                                                                                          
 60       last = first;                                                                                                                                                                                                                                       
 61       first = UINT64_MAX;                                                                                                                                                                                                                                 
 62     }                                                                                                                                                                                                                                                     
 63     if (first != UINT64_MAX && last < first) {                                                                                                                                                                                                            
 64       ranges.clear();                                                                                                                                                                                                                                     
 65       break;                                                                                                                                                                                                                                              
 66     }                                                                                                                                                                                                                                                     
 67     if (!range.empty() && !absl::ConsumePrefix(&range, ",")) {                                                                                                                                                                                            
 68       ENVOY_LOG(debug, "Unexpected characters at the end of range header.");                                                                                                                                                                              
 69       ranges.clear();                                                                                                                                                                                                                                     
 70       break;                                                                                                                                                                                                                                              
 71     }                                                                                                                                                                                                                                                     
 72                                                                                                                                                                                                                                                           
 73     ranges.push_back(RawByteRange(first.value(), last.value()));                                                                                                                                                                                          
 74   }  ```

At the very least, it's not worse, so I'll go with the new version.